client/go/dogma/watch_service.go

```diff
@@ -0,0 +1,378 @@
+	listenersMutex     *sync.Mutex
+
+	doWatchFunc          func(lastKnownRevision int) <-chan *WatchResult
+	convertingResultFunc func(result *WatchResult) *Latest
+
+	projectName string
+	repoName    string
+	pathPattern string
+}
+
+// Latest represents a holder of the latest known value and its Revision retrieved by Watcher.
+type Latest struct {
+	Revision int
+	Value    interface{}
+	Err      error
+}
+
+func newWatcher(projectName, repoName, pathPattern string) *Watcher {
+	rand.Seed(time.Now().UTC().UnixNano())
+	scheduleCTX, scheduleCancelFunc := context.WithCancel(context.Background())
+	watchCTX, watchCancelFunc := context.WithCancel(context.Background())
+	return &Watcher{state: initial, initialChan: make(chan *Latest, 2),
+		scheduleCTX: scheduleCTX, scheduleCancelFunc: scheduleCancelFunc,
+		watchCTX: watchCTX, watchCancelFunc: watchCancelFunc,
+		listenersMutex: &sync.Mutex{},
+		projectName:    projectName,
+		repoName:       repoName, pathPattern: pathPattern}
+}
+
+// AwaitInitialValue awaits for the initial value to be available.
+func (w *Watcher) AwaitInitialValue() Latest {
+	latest := <-w.initialChan
+	// Put it back to the channel so that this can return the value multiple times.
+	w.initialChan <- latest
+	return *latest
+}
+
+// AwaitInitialValue awaits for the initial value to be available during the specified timeout.
+func (w *Watcher) AwaitInitialValueWith(timeout time.Duration) Latest {
+	select {
+	case latest := <-w.initialChan:
+		// Put it back to the channel so that this can return the value multiple times.
+		w.initialChan <- latest
+		return *latest
+	case <-time.After(timeout):
+		return Latest{Err: fmt.Errorf("failed to get the initial value. timeout: %v", timeout)}
+	}
+}
+
+// Latest returns the latest Revision and value of WatchFile() or WatchRepository() result.
+func (w *Watcher) Latest() Latest {
+	if latest := (*Latest)(atomic.LoadPointer(&w.latestPtr)); latest != nil {
+		return *latest
+	} else {
+		return Latest{Err: errors.New("latest is not set yet")}
+	}
+}
+
+// LatestValue returns the latest value of watchFile() or WatchRepository() result.
+func (w *Watcher) LatestValue() (interface{}, error) {
+	latest := w.Latest()
+	if latest.Err != nil {
+		return nil, latest.Err
+	}
+	return latest.Value, nil
+}
+
+// LatestValue returns the latest value of watchFile() or WatchRepository() result. If it's not available, this
+// returns the defaultValue.
+func (w *Watcher) LatestValueOr(defaultValue interface{}) interface{} {
+	latest := w.Latest()
+	if latest.Err != nil {
+		return defaultValue
+	}
+	return latest.Value
+}
+
+// Close stops watching the file specified in the Query or the pathPattern in the repository.
+func (w *Watcher) Close() {
+	atomic.StoreInt32(&w.state, stopped)
+	select {
+	case latest := <-w.initialChan:
+		w.initialChan <- latest
+	case <-time.After(100 * time.Millisecond):
+		w.initialChan <- &Latest{Err: errors.New("watcher is closed")}
```

Comment:
What's the rationale for 100ms timeout? Shouldn't we just make sure no `&Latest` will be sent after `&Latest{Error: ...}` is sent with some concurrency construct?